### DISCUSSION TITLE
Route definition chaining

### DISCUSSION THREAD

**Discussion by nazmulpcc** [clue/framework-x#123] [Ideas]
Currently, a middleware can be added to a route by putting it before the `action` parameter. From the official documentation,

```php
$app->get('/user', new DemoMiddleware(), new UserController());
```
While this works, among other things, if you have many routes, you will end up in a situation where you have some routes that have the action parameter at the second, some has it at the third position.

```php
$app->get('/login', new LoginController());
$app->get('/register', new GuestMiddleware(), new LoginController());
$app->get('/user', new DemoMiddleware(), new UserController());
$app->get('/reset', new ResetPasswordController());
```

This can get a bit confusing. Instead, I think a better syntax would be,

```php
$app->get('/login', new LoginController());
$app->get('/register', new LoginController())
    ->middleware(new GuestMiddleware);
$app->get('/user', new UserController())
    ->middleware(new DemoMiddleware());
$app->get('/reset', new ResetPasswordController());
```

This way, the action comes at a consistent position and it is super clear that we have added a middleware. This syntaxt will also make future improvements easier to add. For example,

```php
$app->get('/login', new LoginController())
    ->name('auth.login') // name the router for referencing it later in the application
    ->limit(1, 5); // limit to 5 requests in 1 minute
```

Let me know what you guys think.

**Comment by SimonFrings** [comment]
@nazmulpcc thanks for bringing this up :+1:

@clue and I also talked about this topic when we were introducing middleware to Framework X. There is no right or wrong with this, both ways work fine, we originally chose the current implementation because it made more sense to us and it also resembles the order of execution (and it is called **middle**ware so it belongs in the middle ^^). I can see why you would consider this approach to be inconsistent and like I said before I don't have anything against your approach, it is as right as ours is.

I don't want to decide on this yet, I rather want to hear what others have to say about this. 🍻

**Reply by nazmulpcc** [reply]
That makes sense to me too. Let's see if anyone else has anything else to say. Personally I think, In a framework like this, having different options makes it easier to onboard different types of people.
